<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen" />
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen" />
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print" />

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>TabAnno by zhanxw</title>
  </head>

  <body>
    <header>
      <div class="inner">
        <h1>TabAnno</h1>
        <h2>TabAnno is an efficient variant annotation tool</h2>
        <a href="https://github.com/zhanxw/anno" class="button"><small>View project on</small>GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <h2>
<a name="contents" class="anchor" href="#contents"><span class="octicon octicon-link"></span></a>Contents</h2>

<p><a href="#Introduction">Introduction</a></p>

<p><a href="#Quick_tutorial">Quick tutorial</a></p>

<p><a href="#Where_to_Find_It">Where to Find It</a></p>

<p><a href="#Usage">Usage</a></p>

<ul>
<li>  <a href="#Command_line">Command line</a>
</li>
<li>  <a href="#Input_files">Input files</a>
</li>
<li>  <a href="#Parameters">Parameters</a>
</li>
</ul><p><a href="#Example">Example</a></p>

<ul>
<li>  <a href="#Built-in_example">Built-in example</a>
</li>
</ul><p><a href="#Contact">Contact</a></p>

<hr><h1>
<a name="introduction" class="anchor" href="#introduction"><span class="octicon octicon-link"></span></a>Introduction</h1>

<p>Anno is short for "annotation", it is used for annotate variants. Our
goal is to provide abundant information for genetic variants promptly.
Some annotation software output only one type of annotation, while Anno
will provide annotations to all transcripts of a gene. Anno support
various file format: VCF file, plain file, plink association output file
and
<a href="http://genome.sph.umich.edu/wiki/EPACTS" title="http://genome.sph.umich.edu/wiki/EPACTS">epacts</a>
file.</p>

<h1>
<a name="quick-tutorial" class="anchor" href="#quick-tutorial"><span class="octicon octicon-link"></span></a>Quick tutorial</h1>

<p>You have an input file in VCF format, and your goal is to annotate it
using refFlat genes database. Then you just need the following command:</p>

<pre><code> anno -i input.vcf -o output.vcf -r hs37d5.fa -g refFlat_hg19.txt.gz -p priority.txt -c codon.txt
</code></pre>

<p>Required files:</p>

<ul>
<li><p><a href="http://www.sph.umich.edu/csg/zhanxw/software/anno/resources/hs37d5.fa" title="http://www.sph.umich.edu/csg/zhanxw/software/anno/resources/hs37d5.fa">hs37d5.fa</a>
Reference genome in NCBI build 37 (you also need to download
<a href="http://www.sph.umich.edu/csg/zhanxw/software/anno/resources/hs37d5.fa.fai" title="http://www.sph.umich.edu/csg/zhanxw/software/anno/resources/hs37d5.fa.fai">hs37d5.fai</a>
)</p></li>
<li><p><a href="http://www.sph.umich.edu/csg/zhanxw/software/anno/resources/refFlat_hg19.txt.gz" title="http://www.sph.umich.edu/csg/zhanxw/software/anno/resources/refFlat_hg19.txt.gz">refFlag_hg19.txt.gz</a>
Gene database in refFlat format (from UCSC website). You can also
use <a href="http://www.sph.umich.edu/csg/zhanxw/software/anno/resources/refFlat.gencode.v7.gz" title="http://www.sph.umich.edu/csg/zhanxw/software/anno/resources/refFlat.gencode.v7.gz">Gencode version
7</a>
or <a href="http://www.sph.umich.edu/csg/zhanxw/software/anno/resources/refFlat.gencode.v11.gz" title="http://www.sph.umich.edu/csg/zhanxw/software/anno/resources/refFlat.gencode.v11.gz">Gencode version
11</a>.</p></li>
<li><p><a href="http://www.sph.umich.edu/csg/zhanxw/software/anno/codon.txt" title="http://www.sph.umich.edu/csg/zhanxw/software/anno/codon.txt">codon.txt</a>
Human codon table.</p></li>
<li><p><a href="http://www.sph.umich.edu/csg/zhanxw/software/anno/priority.txt" title="http://www.sph.umich.edu/csg/zhanxw/software/anno/priority.txt">priority.txt</a>
Priority file, which determines wich annotation type is more
important</p></li>
</ul><p>Outputs:</p>

<p>Annotated VCF will be named <em>output.vcf</em>. Annotations comes in two tags:
ANNO and ANNOFULL. The first tag ANNO showed the most important
annotation types (determined by priority file); the second tag ANNOFULL
showed detailed annotations. Let's see one example annotation:</p>

<pre><code>ANNO=Nonsynonymous:GENE1|GENE3;ANNOFULL=GENE1/CODING_GENE:+:Nonsynonymous(CCT/Pro/P-&gt;CAT/His/H:Base3/30:Codon1/10:Exon1/5):Normal_Splice_Site:Exon|GENE3/CODING_GENE:-:Nonsynonymous(AGG/Arg/R-&gt;ATG/Met/M:Base30/30:Codon10/10:Exon5/5):Normal_Splice_Site:Exon|GENE2/NON_CODING_GENE:+:Upstream
</code></pre>

<p>ANNO tag displayed the most important variant type is <em>Nonsynonymous</em>
and that happened at GENE1 and GENE3; ANNOFULL tag is the full set of
annotation and it consists of two parts, for GENE1 and GENE3
respectively. The first part is for GENE1:</p>

<pre><code>GENE1/CODING_GENE:+:Nonsynonymous(CCT/Pro/P-&gt;CAT/His/H:Base3/30:Codon1/10:Exon1/5):Normal_Splice_Site:Exon
</code></pre>

<p>The format can be explained by sections, and we use annotation for GENE1
as an example:</p>

<ul>
<li>  GENE1 : gene name</li>
<li>  CODING_GENE : transcript name</li>
<li>  <em>+</em> : forward strand</li>
<li>  Nonsynonymous, Normal_Splice_Site, Exon : various annotation types</li>
<li>  CCT/Pro/P-&gt;CAT/His/H : Proline to Histidine mutation</li>
<li>  Base3/30 : mutation happens on the 3rd base of the total 10 bases</li>
<li>  Codon1/10 : mutation happens on the 1st codon of the total 10 codons</li>
<li>  Exon1/5 : mutation happens on the 1st exon of the total 5 exons</li>
</ul><p>Console information:</p>

<p>When anno finished annotation successfully, you will see it output
necessary information as below. You will find basic summary of reference
genome, ts/tv ratio, total number of variants and so on.</p>

<pre><code> The following parameters are available.  Ones with "[]" are in effect:

 Available Options
     Required Parameters : -i [input.vcf], -o [output.vcf]
        Gene Annotations :
                           -g [/net/fantasia/home/zhanxw/anno/resources/refFlat_hg19.txt.gz]
                           -r [/net/fantasia/home/zhanxw/anno/resources/hs37d5.fa]
                           --inputFormat [], --checkReference, -f [], -p []
                           -c [], -u [], -d [], --se [], --si []
       Other Annotations : --genomeScore [], --bed [], --tabix []
     Auxillary Functions : --outputFormat [], --indexOutput
 Use default priority file: /net/fantasia/home/zhanxw/anno/priority.txt
 Use default codon file: /net/fantasia/home/zhanxw/anno/codon.txt
 Load reference genome /net/fantasia/home/zhanxw/anno/resources/hs37d5.fa...
 DONE: 86 chromosomes and 3137454505 bases are loaded.
 Load codon file /net/fantasia/home/zhanxw/anno/codon.txt...
 DONE: codon file loaded.
 Load priority file /net/fantasia/home/zhanxw/anno/priority.txt...
 DONE: 25 priority annotation types loaded.
 Load gene file /net/fantasia/home/zhanxw/anno/resources/refFlat_hg19.txt.gz...
 DONE: 39173 gene loaded.
 DONE: Generated frequency of each annotype type in [ output.vcf.anno.frq ].
 DONE: Generated frequency of each highest priority annotation type in [ output.vcf.top.anno.frq ].
 Ts/Tv ratio: 0.666667
 Ts observed: 4  times; Tv observed: 6 times.
 DONE: Generated frequency of each base change in [ output.vcf.base.frq ].
 DONE: Generated frequency of each codon change in [ output.vcf.codon.frq ].
 DONE: Generated frequency of indel length in [ output.vcf.indel.frq ].
 DONE: 59 varaints are annotated.
 DONE: Generated annotation output in [ output.vcf ].
 Annotation succeed!
</code></pre>

<h1>
<a name="where-to-find-it" class="anchor" href="#where-to-find-it"><span class="octicon octicon-link"></span></a>Where to Find It</h1>

<p>Anno code is hosted online
<a href="https://github.com/zhanxw/anno" title="https://github.com/zhanxw/anno">anno</a>.
You can download the source and compile (type 'make release').</p>

<p>For CSG internal users, the compiled executable file is at:
/net/fantasia/home/zhanxw/anno/executable/anno</p>

<p>The source code is located at:/net/fantasia/home/zhanxw/anno You can
type 'make release' to compile your own executable file. Type "make
test1" or "make test2" will demonstrate the command line to annotate
example VCF files.</p>

<h1>
<a name="usage" class="anchor" href="#usage"><span class="octicon octicon-link"></span></a>Usage</h1>

<h2>
<a name="command-line" class="anchor" href="#command-line"><span class="octicon octicon-link"></span></a>Command line</h2>

<p>After you obtain the anno executable (either by compiling the source
code or by downloading the pre-compiled binary file), you will find the
executable file under executable/anno.</p>

<p>Here is the anno help page by invoking anno without any command line
arguments:</p>

<pre><code> some_linux_host &gt; executable/anno
..............................................       
 ...      G(ene) A(nnotation)                ...     
  ...      Xiaowei Zhan, Goncalo Abecasis     ...    
   ...      zhanxw@umich.edu                    ...  
    ...      Dec 2012                            ... 
     ................................................

Required Parameters
                       -i : Specify input VCF file
                       -o : Specify output VCF file
Gene Annotation Parameters
                       -g : Specify gene file
                       -r : Specify reference genome position
            --inputFormat : Specify format (default: vcf). "-f plain " will use fir
                            st 4 columns as chrom, pos, ref, alt
         --checkReference : Check whether reference alleles matches genome reference
                       -f : Specify gene file format (default: refFlat, other optio
                            ns: knownGene, refGene)
                       -p : Specify priority of annotations
                       -c : Specify codon file (default: codon.txt)
                       -u : Specify upstream range (default: 50)
                       -d : Specify downstream range (default: 50)
                     --se : Specify splice into extron range (default: 3)
                     --si : Specify splice into intron range (default: 8)
           --outputFormat : Specify predefined annotation words (default or epact)
Other Annotation Tools
            --genomeScore : Specify the folder of genome score (e.g. GERP=dirGerp/,
                            SIFT=dirSift/)
                    --bed : Specify the bed file and tag (e.g. ONTARGET1=a1.bed,ONT
                            ARGET2=a2.bed)
                  --tabix : Specify the tabix file and tag (e.g. abc.txt.gz(chrom=1
                            ,pos=7,ref=3,alt=4,SIFT=7,PolyPhen=10)
Please specify input file
</code></pre>

<h2>
<a name="input-files" class="anchor" href="#input-files"><span class="octicon octicon-link"></span></a>Input files</h2>

<p>anno runs on the input VCF file specified on the command-line using flag
'-i'.</p>

<p>Additionally, you need to specify gene file using flag '-g'. You can use
the default refFlat file (using HG19 genome build):
/net/fantasia/home/zhanxw/anno/refFlat_hg19.txt.gz</p>

<h2>
<a name="parameters" class="anchor" href="#parameters"><span class="octicon octicon-link"></span></a>Parameters</h2>

<p>Some of the command line parameters are described here, but most are
self explanatory.</p>

<ul>
<li>  Reference genome file</li>
</ul><pre><code>-r Specify a FASTA format reference genome file.
</code></pre>

<p>Specify <em>-r</em> option enable anno to give more detailed information, for
example, instead of annotating a variant as exon, it will tell you which
codon, which exon that variant locates, whether its
synonymous/non-synonymous and etc.</p>

<p>Anno requires Fasta index file and that will save running memory and
speed up annotation. You can use "samtools faidx XXX.fa" to generate
Fasta index file.</p>

<p>For example, you can specify Fasta file of the whole genome and use "-r
/data/local/ref/karma.ref/human.g1k.v37.fa"</p>

<ul>
<li>  Gene file format</li>
</ul><p>Currently, anno support gene file in refFlat format. A prepared list of
all gene obtained of UCSC website is:
"/net/fantasia/home/zhanxw/anno/refFlat_hg19.txt.gz" . To use that
file, you use flag <em>-g</em>.</p>

<p>As anno support refFlat file by default, you can use refFlat format
without specify gene format flag <em>-f</em>.</p>

<p>To use knownGene or refGene format, you need to specify both <em>-g</em> and
<em>-f</em> flat to tell anno which gene file and which format it is. For
example, <em>-g /net/fantasia/home/zhanxw/anno/knownGene.txt.gz -f
knownGene</em>.</p>

<ul>
<li>  Codon file</li>
</ul><p>Codon file can tell the relationship between triplet codons and amino
acids. A default file is located in:
<em>/net/fantasia/home/zhanxw/anno/codon.txt</em>. If you have special codon
file, you can specify that using flag <em>-c</em>, otherwise, anno will use the
default codon file:</p>

<p><em>default codon file</em></p>

<pre><code># DNA codon table. Stop codon is coded as: Stp, O, Stop
#Codon      AminoAcid   Letter   FullName
AAA     Lys     K       Lysine
AAC     Asn     N       Asparagine
AAG     Lys     K       Lysine
AAT     Asn     N       Asparagine
ACA     Thr     T       Threonine
...
</code></pre>

<ul>
<li>  Annotation ranges</li>
</ul><pre><code>-u how far apart from 5'-end of the gene are counted as upstream
-d how far apart from 3'-end of the gene are counted as downstream
</code></pre>

<p>The <em>-u</em> and <em>-d</em> parameters define the range of upstream and
downstream.</p>

<pre><code>-se how far apart from 5'-end of the gene are counted as upstream
-si how far apart from 3'-end of the gene are counted as upstream
</code></pre>

<p>The <em>-se</em> and <em>-si</em> defines the splice region. By default, 3 bases into
the exon and 8 bases into the introns are defined as splicing sites. If
mutations happen in these regions, we will annotate it as
"Normal_Splice_Site" unless the mutations happens in the traditionally
important "GU...AG" region in the intron, in that case, we will
"Essential_Splice_Site" as annotation.</p>

<ul>
<li>  Annotate by range</li>
</ul><pre><code>                    --bed : Specify the bed file and tag (e.g. ONTARGET1=a1.bed,ONTARGET2=a2.bed)
</code></pre>

<p>BED file is commonly used to represent range. Here you will check if
certain genome position is contained in one or more ranges specified in
the BED format. An example BED file, example.bed, is as follows:</p>

<pre><code>1 10 20 a
1 15 40 b
</code></pre>

<p>There are two ranges: range a is from chromsome 1 position 10
(inclusive) to chromosome 1 position 20 (exclusive); range b is from
chromsome 1 position 15 (inclusive) to chromosome 1 position 40
(exclusive).</p>

<p>When use <em>--bed ONTARGET=example.bed</em> as parameter, the output may look
like this: <em>ONTARGET=a,b</em> in the VCF INFO field.</p>

<ul>
<li>  Annotate by genome score</li>
</ul><pre><code>            --genomeScore : Specify the folder of genome score (e.g. GERP=dirGerp/,SIFT=dirSift/)
</code></pre>

<p>Genome score is a special binary file format (designed by Hyun) storing
per-position genome score, such as GERP score, SIFT score. You will need
to pre-process those scores in a directory. To annotate genome score,
use <em>--genomeScore TAG_NAME=DIRECTORY</em>. In VCF output file, you will
have <em>TAG_NAME=some_value</em> in the INFO field. In other output format,
you will see a separate column.</p>

<ul>
<li>  Annotate by tabix input</li>
</ul><pre><code>                  --tabix : Specify the tabix file and tag (e.g. abc.txt.gz(chrom=1,pos=7,ref=3,alt=4,SIFT=7,PolyPhen=10)
</code></pre>

<p>Tabix file can be used in annotation. Here we required you provide
bgzipped, indexed file. In above example, we use abc.txt.gz as input,
and column 1,7,3,4 as chromosome, position, reference allele,
alternative allele. We also use column 7 as SIFT score and column 10 as
PolyPhen score. Similar syntax as previous, you will get outputs similar
to 'SIFT=0.110;PolyPhen=0.00' in the VCF INFO field. NOTE, in bash,
please use quote around parenthesis, otherwise parenthesis is not
correctly treated. A working example is as follows:</p>

<pre><code>--tabix '/net/fantasia/home/hmkang/bin/annovar/humandb/hg19_ljb_all.txt.gz(chrom=1,pos=2,ref=4,alt=5,mySift=6,mySC=7,myPP2=8,myPC=9)' 
</code></pre>

<h1>
<a name="example" class="anchor" href="#example"><span class="octicon octicon-link"></span></a>Example</h1>

<p>anno can annotate a VCF file and also output statistics of 4 frequency
table: annotation type; base change; codon change; indel size. More
details will be given below.</p>

<h2>
<a name="built-in-example" class="anchor" href="#built-in-example"><span class="octicon octicon-link"></span></a>Built-in example</h2>

<p>In example/ folder, you can see test.vcf, which is a toy example. You
can invoke anno using the following command line:</p>

<pre><code>cd example; ./anno -i test.vcf -r test.fa -g test.gene.txt -c ../codon.txt -o test.out.vcf
</code></pre>

<p>Sample outputs are listed below:</p>

<p>\1) Annotated VCF file, <em>test.out.vcf</em></p>

<pre><code>#VCF_test
#from http://www.sph.umich.edu/csg/liyanmin/vcfCodingSnps/Tutorial.shtml
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  NA12891 NA12892 NA12878
1       3       .       A       G       50      depth=20        ANNO=GENE1/CODING_GENE:+:Exon:Utr5:Normal_Splice_Site|GENE3/CODING_GENE:-:Exon:Utr3:Normal_Splice_Site|GENE2/NON_CODING_GENE:+:Upstream GT:GQ:GD        1/0:31:12       0/0:28:14
1       5       .       C       A       50      depth=20        ANNO=GENE1/CODING_GENE:+:Exon:Nonsynonymous(CCT/Pro/P-&gt;CAT/His/H:Base3/30:Codon1/10:Exon1/5):Normal_Splice_Site|GENE3/CODING_GENE:-:Exon:Nonsynonymous(AGG/Arg/R-&gt;ATG/Met/M:Base30/30:Codon10/10:Exon5/5):Normal_Splice_Site|GENE2/NON_CODING_GENE:+:Upstream   GT:GQ:GD        1/0:31:12       0/0:28:14
...
</code></pre>

<p>The annotation results are stored in the INFO column after ANN= tag. The
annotation format is defined as following:</p>

<p>"|" separates different transcripts, e.g. in the first line, chromosome
1 position 3, there are 3 annotations:
"GENE1/CODING_GENE:+:Exon:Utr5:Normal_Splice_Site" and
"GENE3/CODING_GENE:-:Exon:Utr3:Normal_Splice_Site" and
"GENE2/NON_CODING_GENE:+:Upstream"</p>

<p>":" separates within gene annotation in the following order: gene,
strand, exon/intron, details.</p>

<p>\2) Statistics files:</p>

<p>Four frequency table will be generated after annotation. For example:</p>

<p><em>test.out.vcf.anno.frq</em></p>

<pre><code>Stop_Loss       1
Utr5    2
Utr3    2
CodonRegion     2
CodonGain       2
Frameshift      2
Synonymous      3
StructuralVariation     3
Noncoding       3
Nonsynonymous   4
Deletion        6
Upstream        6
Insertion       6
Essential_Splice_Site   8 
Downstream      8
Intron  12
Exon    21
Normal_Splice_Site      25
</code></pre>

<p><em>test.out.vcf.base.frq</em></p>

<pre><code>A-&gt;G    1
T-&gt;C    1
T-&gt;G    2
A-&gt;C    2
C-&gt;A    5
</code></pre>

<p><em>test.out.vcf.codon.frq</em></p>

<pre><code>Arg-&gt;Met        1
Pro-&gt;Thr        1
Arg-&gt;Arg        1
Pro-&gt;His        1
Gly-&gt;Gly        1
Pro-&gt;Pro        1
Stp-&gt;Tyr        1
Leu-&gt;Val        1
</code></pre>

<p><em>test.out.vcf.indel.frq</em></p>

<pre><code>1       1
-4      1
3       1
-3      1
</code></pre>

<h1>
<a name="contact" class="anchor" href="#contact"><span class="octicon octicon-link"></span></a>Contact</h1>

<p>Questions and requests should be sent to Xiaowei Zhan
(<a href="mailto:zhanxw@umich.edu" title="mailto:zhanxw@umich.edu">zhanxw@umich.edu</a>)
or Goncalo Abecasis
(<a href="mailto:goncalo@umich.edu" title="mailto:goncalo@umich.edu">goncalo@umich.edu</a>)</p>

<p>Author sincerly appreicate Yanming Li for his wonderful tutorial on gene
annotation software
<a href="http://www.sph.umich.edu/csg/liyanmin/vcfCodingSnps/Tutorial.shtml" title="http://www.sph.umich.edu/csg/liyanmin/vcfCodingSnps/Tutorial.shtml">vcfCodingSnps</a>,
and Hyun Ming Kang for his code related to genome scores and his
consistent suggestions and feedbacks.</p>
        </section>

        <aside id="sidebar">
          <a href="https://github.com/zhanxw/anno/zipball/master" class="button">
            <small>Download</small>
            .zip file
          </a>
          <a href="https://github.com/zhanxw/anno/tarball/master" class="button">
            <small>Download</small>
            .tar.gz file
          </a>

          <p class="repo-owner"><a href="https://github.com/zhanxw/anno"></a> is maintained by <a href="https://github.com/zhanxw">zhanxw</a>.</p>

          <p>This page was generated by <a href="pages.github.com">GitHub Pages</a> using the Architect theme by <a href="http://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>

            <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-21871925-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>